### DISCUSSION TITLE
SQLitePCL on iOS without Xamarin

### DISCUSSION THREAD

**Discussion by varondel** [ericsink/SQLitePCL.raw#545] [Q&A] [answered]
Hello,

We are trying to use SQLite in an application using .NetFramework on iOS devices.

**We are not using Xamarin**.
The application is first build on Unity with IL2CPP (Intermediate language is first build in C++), and then build with XCode.

We have been trying many different approaches. Either the dll are not compiling, or the application has runtime error because it tries to use the wrong SQLite provider and fails to load SQLite.

We are using Microsoft.Data.SQLite to create the connection to our database but maybe this is not the right approach ?
Microsoft.Data.Sqlite is internally calling the **Init method of SQLitePCL.Batteries_V2**. Then the provider is **automatically** chosen inside the Init method.

We understand that the Batteries_v2 should use the "sqlite3" provider in order to use the **sytem-provided SQlite of iOS.** But we are failing to find the right dll of Batteries_v2 to do that.

We are now wondering if we are in a dead end because we are not using Xamarin.
Any insight/help about what we should/could use would be greatly appreciated.

Regards,
Vincent

**Comment by ericsink** [comment]
1. In general, SQLitePCLRaw with Unity is tricky. I think it can work, but I'm not sure.

2. The problem you're describing doesn't sound Unity-specific. I would try switching to Microsoft.Data.SQLite.Core and use a specific SQLitePCLRaw provider package instead of a bundle, and do the initialization yourself by calling SetProvider().

**Reply by bricelam** [reply]
Rough draft of instructions for using it on Unity here: https://github.com/ericsink/SQLitePCL.raw/issues/347#issuecomment-664580388

I should really write a full guide for this somewhere...

**Reply by bricelam** [reply]
With IL2CPP, you also need this: (Or did the last time I tried. It might be smarter now.)

https://github.com/bricelam/SQLiteInUnity/blob/main/Assets/Scripts/link.xml

**Comment by varondel** [comment, marked as answer]
Hello,

Thank you all for your replies. 

We finally managed to make it work on Ipad.
Using the dll inside the bundle "**_SQLitePCLRaw.bundle_sqlite3_**" did the trick. I missed this bundle in the nuget explorer of VisualStudio... it was always proposing bundle green or e_sqlite3 but not sqlite3.

This way we don't need to import e_sqlite3.dll in our project, it's targeting the native SQLite libs of iOS
We also still use Microsoft.Data.SQLite to open the connection to the database. It does all the initialization of the batteries and provider on its own. So it's all good.

@bricelam We added this in our link.xml but I'm not even sure if its necessary.
assembly fullname="SQLitePCLRaw.batteries_v2" preserve="all"

Thanks again,
Have a nice day
